bip-0352.mediawiki

```diff
@@ -0,0 +1,471 @@
+
+This gives Bob an alternative to using BIP32 for managing change, while still allowing him to know which of his unspent outputs were change when recovering his wallet from the master key. The change label needs to remain a secret in order to ensure nobody else can label payments as change.
+
+== Specification ==
+
+We use the following functions and conventions:
+
+* ''outpoint'' (36 bytes): the <code>COutPoint</code> of an input (32-byte txid, least significant byte first || 4-byte vout, least significant byte first)<ref name="why_little_endian">'''Why are outpoints little-endian?''' Despite using big endian throughout the rest of the BIP, outpoints are sorted and hashed matching their transaction serialization, which is little-endian. This allows a wallet to parse a serialized transaction for use in silent payments without needing to re-order the bytes when compute the outpoint hash. Note: despite outpoints being stored and serialized as little-endian, the transaction hash (txid) is always displayed as big-endian.</ref>
+* sort<sub>outpoints</sub>(v): sorts a vector ''v'' of ''outpoints'' in ascending order by doing a byte by byte comparison lexicographically.
+* ser<sub>32</sub>(i): serializes a 32-bit unsigned integer ''i'' as a 4-byte sequence, most significant byte first.
+* ser<sub>256</sub>(p): serializes the integer p as a 32-byte sequence, most significant byte first.
+* ser<sub>P</sub>(P): serializes the coordinate pair P = (x,y) as a byte sequence using SEC1's compressed form: (0x02 or 0x03) || ser<sub>256</sub>(x), where the header byte depends on the parity of the omitted y coordinate.
+
+For everything not defined above, we use the notation from [https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#specification BIP340].
+
+=== Versions ===
+
+This document defines ''Silent Payments v0''. Version is communicated through the address in the same way as Segwit addresses. Future upgrades to silent payments will require a new version. As much as possible, future upgrades should support receiving from older wallets (e.g. a silent payments v0 wallet can send to both v0 and v1 addresses). Any changes that break compatibility with older silent payment versions should be a new BIP.
+
+Future Silent Payments versions will use the following versioning scheme:
+
+{| class="wikitable"
+|-
+!
+!0
+!1
+!2
+!3
+!4
+!5
+!6
+!7
+!Compatibility
+|-
+!+0
+|q||p||z||r||y||9||x||8||rowspan="4" | backwards compatible
+|-
+!+8
+|g||f||2||t||v||d||w||0
+|-
+!+16
+|s||3||j||n||5||4||k||h
+|-
+!+24
+|c||e||6||m||u||a||7|| -
+|}
+
+''v31'' (l) is reserved for a backwards incompatible change, if needed. For ''Silent Payments v0'':
+
+* If the receiver's silent payment address version is:
+** ''v0'': check that the data part is exactly 66-bytes. Otherwise, fail
+** ''v1'' through ''v30'': read the first 66-bytes of the data part and discard the remaining bytes (if any)
+** ''v31'': fail
```

Comment:
Some rationale for this would be nice. What kind of changes could be made in a new version without breaking compatibility with v0?

Reply:
Here are a few (admittedly contrived) examples:

* A new output type gets added to the "Inputs for shared secret derivation" list. This would be a new address version so that senders know they can include the new output type when sending to a v1 address, and v0 senders can also send to the v1 address using the v0 rules.
* A new output type gets added and is meant to replace taproot, so now the v1 is interpreted as "create outputs as <new output type> if possible, otherwise create taproot outputs"
* Some solution for optimistic notifications develops and the receiver adds some metadata to the payload to indicate how they want to receive notifications (nostr, simplex, hand-wavy p2p messaging that doesn't exist yet). A v1 sender would parse that extra metadata and use it, whereas a v0 sender would read the first 66 bytes and ignore the rest.
* Some fancy new key type is added for quantum resistance and the key is added as a third public key to the payload. v1 knows to use the third key, whereas a v0 sender use the first 66 bytes and discards the rest

Reply:
@RubenSomsen would also be great if we could come up with a clean example for the BIP, either as an example or as a footnote.